generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java

```diff
@@ -0,0 +1,169 @@
+import static uk.gov.justice.services.common.log.LoggerConstants.SERVICE_CONTEXT;
+import static uk.gov.justice.services.messaging.logging.LoggerUtils.trace;
+
+import uk.gov.justice.services.common.configuration.ServiceContextNameProvider;
+import uk.gov.justice.services.common.converter.StringToJsonObjectConverter;
+import uk.gov.justice.services.messaging.JsonObjectEnvelopeConverter;
+import uk.gov.justice.services.messaging.Metadata;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.Optional;
+
+import javax.inject.Inject;
+import javax.json.Json;
+import javax.json.JsonObject;
+import javax.json.JsonObjectBuilder;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerRequestFilter;
+import javax.ws.rs.container.ContainerResponseContext;
+import javax.ws.rs.container.ContainerResponseFilter;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.ext.Provider;
+
+import org.apache.commons.io.IOUtils;
+import org.slf4j.Logger;
+import org.slf4j.MDC;
+
+/**
+ * Filter listens to all requests and adds request information to the Logger Mapped Diagnostic
+ * Context.  This can be added to the log output by setting %X{requestData} in the logger pattern.
+ */
+@Provider
+public class LoggerRequestDataFilter implements ContainerRequestFilter, ContainerResponseFilter {
+
+    private static final String JSON_MEDIA_TYPE_FORMAT = "+json";
+    private static final String PAYLOAD_METADATA = "_metadata";
+
+    @Inject
+    Logger logger;
+
+    @Inject
+    ServiceContextNameProvider serviceContextNameProvider;
+
+    private static Metadata payloadMetadataFrom(final String message) {
+        return new JsonObjectEnvelopeConverter()
+                .asEnvelope(new StringToJsonObjectConverter().convert(message))
+                .metadata();
+    }
+
+    @Override
+    public void filter(final ContainerRequestContext requestContext) throws IOException {
+        trace(logger, () -> "Adding request data to MDC");
+
+        final JsonObjectBuilder builder = Json.createObjectBuilder();
+        final MultivaluedMap<String, String> headers = requestContext.getHeaders();
+
+        addServiceContextNameIfPresent(builder);
+
+        Optional.ofNullable(headers.getFirst(CONTENT_TYPE)).ifPresent(value -> builder.add(CONTENT_TYPE, value));
+        Optional.ofNullable(headers.getFirst(ACCEPT)).ifPresent(value -> builder.add(ACCEPT, value));
+
+        mergeHeadersWithPayloadMetadataIfPresent(requestContext, headers)
+                .ifPresent(metadataBuilder -> builder.add(METADATA, metadataBuilder));
+
+        MDC.put(REQUEST_DATA, builder.build().toString());
+
+        trace(logger, () -> "Request data added to MDC");
+    }
+
+    @Override
+    public void filter(final ContainerRequestContext requestContext, final ContainerResponseContext responseContext) throws IOException {
```

Comment:
I have extracted common code, but due to uncommon usage I have avoided abstract class